control_plane/storcon_cli/src/main.rs

```diff
@@ -0,0 +1,643 @@
+    },
+    ServiceRegister {
+        #[arg(long)]
+        http_host: String,
+        #[arg(long)]
+        http_port: u16,
+        #[arg(long)]
+        region_id: String,
+        #[arg(long)]
+        availability_zone_id: String,
+        #[arg(long)]
+        version: u64,
+        // {
+        //   "version": 1,
+        //   "host": "${HOST}",
+        //   "port": 6400,
+        //   "region_id": "{{ console_region_id }}",
+        //   "instance_id": "${INSTANCE_ID}",
+        //   "http_host": "${HOST}",
+        //   "http_port": 9898,
+        //   "active": false,
+        //   "availability_zone_id": "${AZ_ID}",
+        //   "disk_size": ${DISK_SIZE},
+        //   "instance_type": "${INSTANCE_TYPE}",
+        //   "register_reason" : "New pageserver"
+        // }
+    },
+    TenantOnboard {
+        #[arg(long)]
+        tenant_id: TenantId,
+        #[arg(long)]
+        target_pageserver_id: NodeId,
+    },
+    TenantScatter {
+        #[arg(long)]
+        tenant_id: TenantId,
+    },
+    TenantDescribe {
+        #[arg(long)]
+        tenant_id: TenantId,
+    },
+}
+
+/// Request format for control plane POST /management/api/v2/pageservers
+///
+/// This is how physical pageservers register, but in this context it is how we
+/// register the storage controller with the control plane, as a "virtual pageserver"
+#[derive(Serialize, Deserialize, Debug)]
+struct ServiceRegisterRequest {
+    version: u64,
+    host: String,
+    port: u16,
+    /// This is the **Neon** region ID, which looks something like `aws-eu-west-1`
+    region_id: String,
+    /// This expects an EC2 instance ID, for bare metal pageservers.  But it can be any unique identifier.
+    instance_id: String,
+    http_host: String,
+    http_port: u16,
+    /// This is an EC2 AZ name (the ZoneName, not actually the AZ ID).  e.g. eu-west-1b
+    availability_zone_id: String,
+    disk_size: u64,
+    /// EC2 instance type.  If it doesn't make sense, leave it blank.
+    instance_type: String,
+    /// Freeform memo describing the request
+    register_reason: String,
+    // Set to true to indicate that this 'pageserver' is really a storage controller
+    // FIXME: have to omit this because control plane won't allow setting it for existing pageserver
+    is_storage_controller: bool,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct ServiceRegisterResponse {
+    // This is a partial representation of the management API's swagger `Pageserver` type.  Unused
+    // fields are ignored.
+    id: u64,
+    node_id: u64,
+    instance_id: String,
+}
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+#[command(arg_required_else_help(true))]
+struct Cli {
+    #[arg(long)]
+    api: Url,
+
+    #[arg(long)]
+    jwt: Option<String>,
+
+    #[command(subcommand)]
+    command: Command,
+}
+
+struct Client {
+    base_url: Url,
+    jwt_token: Option<String>,
+    client: reqwest::Client,
+}
+
+impl Client {
```

Comment:
nit: Would prefer to have the client in a separate crate, like I did with pagebench . Some arguments:
* It makes this file shorter & focussed on the CLI aspect
* It makes the client ready for re-use (e.g., you re-used pageserver_client in the storage controller server impl IIRC)

I guess the price to pay is that for the client to be a proper abstraction, one has to add wrappers for the `.dispatch()` method.
But, that seems like a price worth paying forward to me.

Anyways, just a suggestion

Reply:
I think that'll be worth doing if we find ourselves re-using any of this: while this is all just a single debug CLI I think the convenience of having a single file outweighs the motivation to separate client from CLI (quite a bit of boilerplate)